app/mempool.go

```diff
@@ -0,0 +1,315 @@
+	// load from config file as a config for mempool
+	txDecoder sdk.TxDecoder
+}
+
+// Insert inserts a tx into mempool, currently only used for rawData from tx related to 2-phases aggregation of oracle modul
+func (em *ImuaMempool) Insert(_ context.Context, tx sdk.Tx) error {
+	// we don't filter tx not with message of rawData type, those tx will just be added into tendermint's txpool
+	if !em.includesMsgOracle(tx) {
+		return nil
+	}
+
+	piece, msgOracle, _, isRawData := em.GetPieceWithProof(tx)
+	// we don't filter tx not with message of rawData type
+	if !isRawData {
+		return nil
+	}
+	if piece == nil {
+		return errors.New("failed to parse pieceWithProof from tx with oracle rawData message")
+	}
+	fID := msgOracle.FeederID
+
+	piecesCached, ok := em.cachedPieces[fID]
+	if !ok {
+		em.cachedPieces[fID] = make(map[uint32][]*oracletypes.PieceWithProof)
+		em.cachedPieces[fID][piece.Index] = []*oracletypes.PieceWithProof{piece}
+		em.count++
+		return nil
+	}
+
+	piecesIndexCached, ok := piecesCached[piece.Index]
+	if !ok {
+		piecesCached[piece.Index] = []*oracletypes.PieceWithProof{piece}
+		em.count++
+		return nil
+	}
+	for _, pieceCached := range piecesIndexCached {
+		if pieceCached.EqualsTo(piece) {
+			return errors.New("piece exists in mempool")
+		}
+	}
+	piecesCached[piece.Index] = append(piecesIndexCached, piece)
+	em.count++
+	return nil
+}
+
+// Select selects txs for block proposal, we only select rawData txs with rawData message which is used for 2-phases aggregation of oracle module, for other txs we just return them
+// we only keep one tx for each feederID, and we only keep the tx with the piece index expected by the feederID
+func (em *ImuaMempool) Select(ctx context.Context, txList [][]byte) mempool.Iterator {
+	// remove all expired tx, when Select for block 100, all txs belongs to 99 or before should be removed
+
+	// feederIDS:[]uint64, which are expecting rawData
+	// []Tx, each feederID must have one tx
+	collectingFeederIDs := em.k.FeederManager.FeederIDsCollectingRawData()
+	if len(collectingFeederIDs) == 0 {
+		// remove all cached pieces since no collectingFeederIDs available
+		em.reset()
+		txDecodedList := make([]sdk.Tx, 0, len(txList))
+		for _, txBytes := range txList {
+			tx, err := em.txDecoder(txBytes)
+			if err != nil {
+				continue
+			}
+			if _, _, isRawData, _ := utils.OracleCreatePriceTx(tx); isRawData {
+				// remove rawData from request txList if there's no feederID collecting rawData pieces
+				continue
+			}
+			txDecodedList = append(txDecodedList, tx)
+		}
+		if len(txDecodedList) > 0 {
+			return IteratorFromSlice(txDecodedList)
+		}
+		return nil
+	}
+
+	// remove all expired txs from cache
+	em.clearExpiredFeederIDcache(collectingFeederIDs)
+
+	seenFeederIDs := make(map[uint64]struct{})
+	keep := make([]sdk.Tx, 0, len(txList))
+	for _, txBytes := range txList {
+		tx, err := em.txDecoder(txBytes)
+		if err != nil {
+			// skip undecodable bytes - they will be recehecked when re-broadcast
```

Comment:
```suggestion
			// skip undecodable bytes - they will be rechecked when re-broadcast
```
Why skip and not drop them? Will they become `decodable` later?

Reply:
The result of decoding a transaction can only differ if a different txDecoder is used. In the default PrepareProposalHandler, any transaction that fails to decode will be caught by h.txVerifier.PrepareProposalVerifyTx, which subsequently calls Remove on the mempool to discard the invalid transaction. Therefore, it’s safe to simply skip processing it here.

Reply:
Got it, please fix the typo at your convenience.

Reply:
fixed